File: LiKitty/SplitModel.m
//
//  SplitModel.m
//  LiKitty
//
//  Created by Jeremy Chen on 4/25/15.
//  Copyright (c) 2015 Jeremy Chen. All rights reserved.
//

#import "SplitModel.h"
#import "Person.h"

@interface SplitModel ()
// private properties
@property (strong, nonatomic) NSMutableArray *people;
@property (strong, nonatomic) NSMutableDictionary *items;

//@property (strong, nonatomic) Item *currentItem;
@property (nonatomic) double taxRate;
@property (nonatomic) double tipRate;

@end

@implementation SplitModel

// Singleton model
+ (instancetype) sharedModel {
    static SplitModel *_sharedModel = nil;
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedModel = [[self alloc] init];
    });
    
    return _sharedModel;
}

// Initialize, with default tax rate of 9% and tip rate of 15%
- (instancetype) init
{
    self = [super init];
    if (self) {
        self.items = [[NSMutableDictionary alloc] init];
        self.people = [[NSMutableArray alloc]init];

        self.taxRate = .08;
        self.tipRate = .15;
    }
    
    return self;
}

// Add item to model with its price and the people involved
- (void) addItemWithName: (NSString*) itemName
               withPrice: (double) itemPrice
              withPeople: (NSMutableArray*) people
             isTaxExempt: (bool) isTaxExempt
{
    
    NSMutableDictionary* itemDict = [[NSMutableDictionary alloc] initWithObjectsAndKeys:
                                     [NSNumber numberWithDouble:itemPrice], @"Price",
                                    @(isTaxExempt), @"TaxExempt",
                                     people, @"People",
                                     nil];
    
    // Add item to items dictionary with its name as key
    [self.items setObject: itemDict forKey: itemName];
}

- (BOOL) checkIfItemExists: (NSString*) itemName {
    if ([self.items objectForKey: itemName]) {
        return TRUE;
    }
    return FALSE;
}

- (NSMutableDictionary*) getItems {
    return self.items;
}

// Delete item based on index in dictionary
- (void) deleteItemAtIndex: (NSUInteger) index {
    NSArray *keys = [self.items allKeys];
    [self.items removeObjectForKey: [keys objectAtIndex: index]];
}

// Get item based on index in dictionary
- (NSString*) getItemNameAtIndex: (NSUInteger) index {
    NSArray *keys = [self.items allKeys];
    return[keys objectAtIndex: index];
}

// Get item's price at index of dictionary
- (NSNumber*) getItemPriceAtIndex: (NSUInteger) index {
    NSArray *keys = [self.items allKeys];
    id keyAtIndex = [keys objectAtIndex: index];
    NSMutableDictionary* itemDict = [self.items objectForKey: keyAtIndex];
    
    return [itemDict objectForKey: @"Price"];
}

// Return number of items in model
- (NSInteger) count {
    return [self.items count];
}

// Add person to the model
- (void) addPerson: (Person*) person {
    [self.people addObject:person];
}

// Return person via name if he/she exists in model. Otherwise return nothing
- (Person*) returnPersonIfExists: (NSString*) name {
    for (id person in self.people) {
        if ([[person getName] isEqualToString: name]) {
            return person;
        }
    }
    return nil;
}

- (NSMutableArray*) getPeople {
    return self.people;
}

// Get subtotal of items (doesn't include tax and tip)
- (double) getSubtotal {
    double subtotal = 0;
    
    // Iterate through each item, totaling its price
    for(id key in self.items) {
        NSMutableDictionary* itemDict = [self.items objectForKey:key];
        subtotal += [[itemDict objectForKey: @"Price"] doubleValue];
    }
    
    return subtotal;
}

- (double) getSubtotalWithoutTaxExempt {
    double subtotal = 0;
    
    // Iterate through each item, totaling its price
    for(id key in self.items) {
        NSMutableDictionary* itemDict = [self.items objectForKey:key];
        if ([[itemDict objectForKey:@"TaxExempt"] isEqualToNumber:[NSNumber numberWithBool:NO]]) {
            subtotal += [[itemDict objectForKey: @"Price"] doubleValue];
        }
    }
    
    return subtotal;
}

- (double) getTaxAmount {
    return self.taxRate * [self getSubtotalWithoutTaxExempt];
}

- (double) getTipAmount {
    return self.tipRate * [self getSubtotal];
}

- (double) getTaxR {
    return self.taxRate;
}

- (double) getTipR {
    return self.tipRate;
}

- (void) setTaxR: (double) taxRate {
    self.taxRate = taxRate;
}

- (void) setTipR: (double) tipRate {
    self.tipRate = tipRate;
}

// Add subtotal with tax and tip
- (double) getTotal {
    return [self getSubtotal] + self.getTaxAmount + self.getTipAmount;
}

// Clear everything, and reset tax and tip rate to default
- (void) resetEverything {
    [self.items removeAllObjects];
    [self.people removeAllObjects];

    self.taxRate = .09;
    self.tipRate = .15;
}

@end

File: LiKitty/SplitModel.swift
//
//  SplitModel.swift
//  LiKitty
//
//  Created by Jeremy Chen on 3/23/16.
//  Copyright © 2016 Jeremy Chen. All rights reserved.
//

import Foundation

class SplitModel {
    static let sharedInstance = SplitModel()
    
    let DEFAULT_TAX_RATE = 0.08
    let DEFAULT_TIP_RATE = 0.15
    
    var savedPeople: [Person]!
    var savedItems: [String:Item]!
    var currentTaxRate: Double
    var currentTipRate: Double
    
    init() {
        print("SplitModel singleton initiated")
        
        currentTaxRate = DEFAULT_TAX_RATE
        currentTipRate = DEFAULT_TIP_RATE
        
    }
    
    /* PEOPLE */
    func addPerson(person: Person) {
        savedPeople.append(person)
    }
    
    func returnPersonByName(name: String) -> Person? {
        for p in savedPeople {
            if (p.getName() == name) {
                return p
            }
        }
        return nil
    }
    
    func returnAllPeople() -> [Person] {
        return savedPeople
    }
    
    /* MONEY */
    func getSubtotal() -> Double {
        
    }
    
    func getSubtotalMinusItemsWithTaxIncluded() -> Double {
        
    }
    
    func getTaxAmount() -> Double {
        
    }
    
    func getTipAmount() -> Double {
        
    }
    
    func getTaxRate() {
        
    }
    
    func getTipRate() {
        
    }
    
    func setTaxRate() {
        
    }
    
    func setTipRate() {
        
    }
    
    func getTotal() -> Double {
        return getSubtotal() + getTaxAmount() + getTipAmount()
    }
    
    /* ITEMS */
    func addItem(itemName: String,
        itemPrice: Double,
        peopleWithPercentage: [(Person,Double)],
        taxIncluded: Bool) -> Bool {
        
        if (createAndSaveItem(itemName, itemPrice: itemPrice)) {
            // For each person, add item with percentage
            for p in peopleWithPercentage {
                let person = p.0
                let percentage = p.1
                
                person.addFood(itemName, withPrice: itemPrice, withContribution: percentage)
            }
            
            return true
        }
        
        return false
    }
    
    // Returns all items
    func getAllItems() -> [String:Item] {
        return savedItems
    }
    
    // Return all item names
    func getAllItemNames() -> [String] {
        return [String](savedItems.keys)
    }
    
    // Returns item by name
    func getItemByName(itemName: String) -> Item {
        return savedItems[itemName]!
    }
    
    // Return number of items
    func getItemCount() -> Int {
        return savedItems.count
    }
    
    // Deletes item by name
    func deleteItem(itemName: String) {
        savedItems.removeValueForKey(itemName)
    }
    
    private func createAndSaveItem(itemName: String, itemPrice: Double) -> Bool {
        if (savedItems[itemName] != nil) {
            return false
        }
        
        let item = Item(itemName: itemName, itemPrice: itemPrice)
        savedItems[itemName] = item
        
        return true
    }
    
    
    
    
    
    
    // Save items and people into CoreData
    func save() {
        
    }
    

}













